### PR TITLE
JS2 HIR Implementation

### PR DESCRIPTION
**Pull request by ambiguousname**
Per #420, this is my ongoing effort to implement the HIR backend for Javascript and Typescript.

There's still quite a bit to be done (lots of backend functions with TODOs, entry-level matches for hir types, etc.), but this is for anyone to get a good look to see if there's any issues with the way that I'm approaching things.

This is mostly based on the Dart backend, with some added comments that were for my own understanding during development.

Let me know if there's anything I can clarify!

### CODE REVIEW

File: core/src/hir/attrs.rs [hunk 1]
```diff
@@ -596,21 +596,116 @@ impl Attrs {
...
+    /// I.E.
+    /// ```dart
+    /// class Iterator {
+    ///     int _iteratorNext()
+    /// }
+    /// ```
     pub iterators: bool,
+    /// I.E.
+    /// ```dart
+    /// class Sample {
```
Comment (robertbastian) [inline core/src/hir/attrs.rs:698]:
```suggestion
    /// class Sample implements Iterable {
```

File: core/src/hir/attrs.rs [hunk 2]
```diff
@@ -596,21 +596,116 @@ impl Attrs {
...
+    /// ```cpp
+    /// HeapItem* item = new HeapItem();
+    /// ```
+    /// Would be an item that supports memory sharing.
     pub memory_sharing: bool,
+    /// I.E.
+    /// ```dart
+    /// class Iterator {
+    ///     int _iteratorNext()
+    /// }
```
Comment (robertbastian) [inline core/src/hir/attrs.rs:693]:
```suggestion
    /// class MyIterator implements Iterator {
    ///     int next()
    /// }
```

File: core/src/hir/attrs.rs [hunk 3]
```diff
@@ -596,21 +596,116 @@ impl Attrs {
...
+    /// class Sample {
+    ///     @override
+    ///     String toString()
+    /// }
+    /// ```
     pub stringifiers: bool,
+    /// I.E.
+    /// ```dart
+    /// class Sample {
+    ///     @override bool operator ==(Object other) => other is {{type_name}} && compareTo(other) == 0;
```
Comment (robertbastian) [inline core/src/hir/attrs.rs:678]:
```suggestion
    ///     @override bool operator ==(Object other) => ...;
```

File: core/src/hir/attrs.rs [hunk 4]
```diff
@@ -596,21 +596,116 @@ impl Attrs {
...
+    /// }
+    /// ```
     pub comparators: bool,
+    /// Can items with `#[diplomat::attr(supports= memory_sharing)]` be shared across threads?
```
Comment (robertbastian) [inline core/src/hir/attrs.rs:682]: I think this is dead code, @Manishearth?

File: core/src/hir/methods.rs [hunk 5]
```diff
@@ -31,8 +31,12 @@ pub struct Method {
 #[derive(Debug, Clone)]
 #[non_exhaustive]
 pub enum SuccessType {
+    /// Is the return type something that can be written to? I.e., a buffer
```
Comment (robertbastian) [inline core/src/hir/methods.rs:34]:
```suggestion
    /// Conceptually returns a string, which gets written to the `write: DiplomatWrite` argument
```

File: core/src/hir/methods.rs [hunk 6]
```diff
@@ -31,8 +31,12 @@ pub struct Method {
...
     Write,
+    /// See [`OutType`] and [`super::ty_position::TyPosition`].
+    /// Basically, it's a Box<SomeType> that we can output, but not recieve again as input.
```
Comment (robertbastian) [inline core/src/hir/methods.rs:37]:
```suggestion
    /// A Diplomat type. Some types can be outputs, but not inputs, which is expressed by the `OutType` parameter.
```

File: feature_tests/js2/api/AttrEnum.d.ts
Comment (robertbastian) [inline feature_tests/js2/api/AttrEnum.d.ts:1]: nit: use `.gitattributes` to mark these as generated files

File: newjsbackendnotes.md
Comment (robertbastian) [inline newjsbackendnotes.md:1]: nit: keep this outside the repo

File: support/functions.ds
Comment (robertbastian) [inline support/functions.ds:1]: suggestion: merge all windows fixes as a separate PR

File: tool/src/lib.rs [hunk 7]
```diff
@@ -75,6 +77,41 @@ pub fn gen(
...
+            let mut attr_validator = hir::BasicAttributeValidator::new("js2");
+            attr_validator.other_backend_names.push("js".into());
+
+            attr_validator.support.renaming = true;
+            attr_validator.support.disabling = true;
+
+            attr_validator.support.iterables = true;
+            attr_validator.support.iterators = true;
+            
+            attr_validator.support.accessors = true;
```
Comment (robertbastian) [inline tool/src/lib.rs:90]: I'd disable these for now, special method support can be added incrementally

File: tool/templates/js2/runtime.d.ts
Comment (robertbastian) [inline tool/templates/js2/runtime.d.ts:1]: I assume this is duplicated in `src/js`? Dedupe

File: tool/templates/js2/runtime.mjs [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	wasm.diplomat_buffer_writeable_destroy(writeable);
+	}
+}
+
+export class FFIError extends Error {
+	constructor(error_value) {
+	super("Error over FFI");
+	this.error_value = error_value; // (2)
+	}
+}
```
Comment (robertbastian) [inline tool/templates/js2/runtime.mjs:28]: note: I would like to get rid of this error wrapping in JS2. If a type is used in the error position of a `Result<T, E>`, it it should implement `Error` itself.

File: tool/templates/js2/runtime.mjs
Comment (robertbastian) [inline tool/templates/js2/runtime.mjs:1]: also duplicate

**Review by robertbastian — COMMENTED** [review]
Preliminary review, haven't looked in-depth at the generated code yet

File: tool/templates/js2/runtime.d.ts
Comment (ambiguousname) [reply tool/templates/js2/runtime.d.ts:1]: Yes, the only major difference is that js2/runtime.d.ts defines "pointer" as a type for some clarity when reading .d.ts classes with functions (like constructors) that take a pointer.

Comment (robertbastian) [reply tool/templates/js2/runtime.d.ts:1]: I see, you can add this to the JS1 runtime though, it doesn't hurt there

Comment (ambiguousname) [reply tool/templates/js2/runtime.d.ts:1]: For sure. Would you like me to just delete these from the templates folder and include_str! from the js folder?

Comment (robertbastian) [reply tool/templates/js2/runtime.d.ts:1]: You can delete the JS1 ones and `include!` the new ones there.

File: tool/templates/js2/runtime.mjs (hunk 8 above)
Comment (ambiguousname) [reply tool/templates/js2/runtime.mjs:28]: That would be nice. I think it might require adding information to the HIR for Enums and Structs to see if they're ever used in that position

Comment (ambiguousname) [reply tool/templates/js2/runtime.mjs:28]: Writing this since I just started encountering this with Result<> types, so I'm defaulting to the FFIError solution for now

**Comment by ambiguousname** [comment]
Updated, all feature tests are now passing.

There are still quite a lot of TODOs and FIXMEs that involve things not covered in the feature tests, so I will be coming back to this later to try and fill in the gaps.

**Comment by sffc** [comment]
We should land something even if it's not complete because it allows other people to start contributing any missing pieces

File: example/js2/api/ICU4XFixedDecimal.mjs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        try {
+    
+            return diplomatRuntime.readString8(wasm, wasm.diplomat_buffer_write_get_bytes(write), wasm.diplomat_buffer_write_len(write));
+        } finally {
+        
+            wasm.diplomat_free(diplomat_receive_buffer, 5, 4);
+        
+            wasm.diplomat_buffer_write_destroy(write);
+        
+        }
```
Comment (sffc) [inline example/js2/api/ICU4XFixedDecimal.mjs:75]: Question: why aren't you using `withDiplomatWrite`? (it's okay, but why the change?)

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    // Since JS won't garbage collect until there are no incoming edges.
+    #selfEdge = [];
+    
+    
+    constructor(ptr, selfEdge) {
+        
+        this.#ptr = ptr;
+        this.#selfEdge = selfEdge;
+        // Unconditionally register to destroy when this object is ready to garbage collect.
+        ICU4XFixedDecimalFormatter_box_destroy_registry.register(this, this.#ptr);
```
Comment (sffc) [inline example/js2/api/ICU4XFixedDecimalFormatter.mjs:36]: Question: Is this part of the code generated conditionally on whether the type is owned? I see previously it did a runtime dynamic check on an `owned` boolean.

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    
+        try {
+    
+            if (!diplomatRuntime.resultFlag(wasm, diplomat_receive_buffer, 4)) {
+                throw diplomatRuntime.FFIError(null);
+            }
+            return new ICU4XFixedDecimalFormatter(diplomatRuntime.ptrRead(wasm, diplomat_receive_buffer), []);
+        } finally {
+        
+            this.free(); /* TODO: Does this work? */
```
Comment (sffc) [inline example/js2/api/ICU4XFixedDecimalFormatter.mjs:57]: Issue: `this.free()` isn't a function in this context as far as I can tell. What are you trying to do?

File: example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        this.#someOtherConfig = value;
+    }
+
+    // Return this struct in FFI function friendly format.
+    // Returns an array that can be expanded with spread syntax (...)
+    
+    _intoFFI() {
+        return [
+            groupingStrategy.ffiValue, 
+            someOtherConfig]
```
Comment (sffc) [inline example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs:31]: Observation: this turns a struct into an ordered list of its fields, which is then spread into separate arguments when passed into the WASM export. Is that correctly aligned with what WASM expects?

**Review by sffc — COMMENTED** [review]
Praise: Good work leveraging the HIR; I still want to give `feature_tests` output a more thorough review.

File: example/js2/api/ICU4XFixedDecimal.mjs (hunk 9 above)
Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimal.mjs:75]: This way fits in much more nicely with the templating system. If you check [method.js.jinja](https://github.com/ambiguousname/diplomat/blob/js2-hir-implementation/tool/templates/js2/method.js.jinja), there are sections for allocation and freeing of whatever is set up to call the FFI function.

withDiplomatWrite() does a lot of the same things but you need to then wrap everything that comes after that function inside the callback. It becomes difficult to include that as a possibility while handling multiple different use cases in the same generic way, for instance: `-> Result<Write, ()>` versus `-> Write` versus `Result<OutType, ()>`, etc.

TL;DR: It's a more generic way of crafting these functions that allows for more flexibility.

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs (hunk 10 above)
Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimalFormatter.mjs:36]: I believe so, I checked in with @robertbastian to see if this was the kind of behavior we wanted for ownership of opaques. I can give it a more thorough look later, however.

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs (hunk 11 above)
Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimalFormatter.mjs:57]: Oh, this random free statement is something that's been since fixed in the wasm-demo-gen branch, I just forgot to run the code on the example file for this branch.

Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimalFormatter.mjs:57]: I believe feature_tests is the more recent version, but I will go and update `example` real quick.

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs (hunk 10 above)
Comment (Manishearth) [reply example/js2/api/ICU4XFixedDecimalFormatter.mjs:36]: Yeah I think in Dart we only register for types that are owned? And all references to them are stashed as edges.

File: example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs (hunk 12 above)
Comment (Manishearth) [reply example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs:31]: IIRC yes? but yes, let's double check that

File: example/js2/api/ICU4XFixedDecimalFormatter.mjs (hunk 11 above)
Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimalFormatter.mjs]: Okay, this file has been updated with the latest commit, and `this.free()` has since been removed.

File: example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs (hunk 12 above)
Comment (ambiguousname) [reply example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs]: I just did, through the LLVM IR output:

```ll
; diplomat_example::decimal::ffi::ICU4XFixedDecimalFormatter::try_new
; Function Attrs: nounwind
define hidden align 4 ptr @_ZN16diplomat_example7decimal3ffi26ICU4XFixedDecimalFormatter7try_new17h1c21e7448442dd44E
(ptr align 4 %locale, ptr align 4 %provider, i32 %options.0, i1 zeroext %options.1)
unnamed_addr #0 !dbg !7776 {
```

`i32 %options.0` is most likely `grouping_strategy` and `i1 %options.1` is most likely `some_other_config`.

File: tool/src/js2/formatter.rs [hunk 13]
```diff
@@ -0,0 +1,238 @@
+use std::borrow::Cow;
```
Comment (Manishearth) [inline tool/src/js2/formatter.rs:1]: nit: make sure to run `cargo fmt` on this code

**Review by Manishearth — COMMENTED** [review]
code looks roughly right. I'll have a look at the lifetime stuff eventually but we shouldn't block this on that

File: newjsbackendnotes.md [hunk 14]
```diff
@@ -0,0 +1,5 @@
+- Want to use HIR
```
[the remark this answers is not in the document]
Comment (ambiguousname) [reply newjsbackendnotes.md:1]: Oh, I totally meant to remove this once the pull request was open. I'll go ahead and delete it

**Comment by ambiguousname** [comment]
The notes doc I just removed had things that were mostly out of date. The only thing that I'd like to re-iterate here is that I make sure to test the build scripts on Node v20

File: example/js2/tsconfig.json
Comment (robertbastian) [inline example/js2/tsconfig.json:1]: If you've got TS tests to work, please commit them in a separate PR to JS1 as well (#504 )

**Review by robertbastian — COMMENTED** [review]
Not a thorough review, but posting my comments so far.

Re Node 20, I merged #503, which tests Node 18, 20, and 22. This also includes the `wasm.mjs` change, so if you merge that will disappear from this PR.

File: support/functions.ds [hunk 15]
```diff
@@ -67,7 +67,7 @@ fn generate_generic
     rm -r "${backend}/${output_folder}${preserve_subdirectory}"
     # we do a /.. so that cases where output_folder is multiple folders down
     # end up copying to the right path
-    mv "tmp/${output_folder}${preserve_subdirectory}" "${backend}/${output_folder}${preserve_subdirectory}/.."
+    mv "tmp/${output_folder}${preserve_subdirectory}/" "${backend}/${output_folder}${preserve_subdirectory}/.."
```
Comment (robertbastian) [inline support/functions.ds:70]: this probably fixes https://github.com/rust-diplomat/diplomat/issues/440, please put in a separate PR

File: ContributingOnWindows.md
[the remark this answers is not in the document]
Comment (ambiguousname) [reply ContributingOnWindows.md:1]: https://github.com/rust-diplomat/diplomat/pull/507

File: support/functions.ds (hunk 15 above)
Comment (ambiguousname) [reply support/functions.ds:70]: https://github.com/rust-diplomat/diplomat/pull/508

File: example/js2/tsconfig.json
Comment (ambiguousname) [reply example/js2/tsconfig.json:1]: I configured the tests to compile correctly, but they don't assert the correct types or anything like that. Still, I'll set up a separate PR for JS1

File: support/functions.ds (hunk 15 above)
Comment (ambiguousname) [reply support/functions.ds:70]: #508

File: example/js2/tsconfig.json
Comment (ambiguousname) [reply example/js2/tsconfig.json:1]: https://github.com/rust-diplomat/diplomat/pull/509

But this isn't a full fix, there are some things that need to be adjusted in the backend as well (more detail in the PR)

**Comment by ambiguousname** [comment]
Note: test coverage for NestedBorrowFields would be good for a lot of features. I'll try to add that

File: feature_tests/js2/api/MyString.mjs [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    getBoxedStr() {
+        
+        const diplomat_receive_buffer = wasm.diplomat_alloc(8, 4);
+        const result = wasm.MyString_get_boxed_str(diplomat_receive_buffer, this.ffiValue);
+    
+        try {
+    
+            return diplomat_receive_buffer // TODO: Slice c_to_js;
```
[the remark this answers is not in the document]
Comment (ambiguousname) [reply feature_tests/js2/api/MyString.mjs]: Done!

File: feature_tests/js2/api/NestedBorrowedFields.mjs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    
+
+    _fromFFI(ptr, xEdges, yEdges, zEdges) {
+        const fieldsDeref = ptr;
+        this.#fields = new BorrowedFields()._fromFFI(fieldsDeref, xEdges);
+        const boundsDeref = ptr;
+        this.#bounds = new BorrowedFieldsWithBounds()._fromFFI(boundsDeref, xEdges, yEdges, yEdges);
+        const bounds2Deref = ptr;
+        this.#bounds2 = new BorrowedFieldsWithBounds()._fromFFI(bounds2Deref, zEdges, zEdges, zEdges);
```
[the remark this answers is not in the document]
Comment (ambiguousname) [reply feature_tests/js2/api/NestedBorrowedFields.mjs]: This is also done.

File: feature_tests/js2/api/BorrowedFields.mjs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    _fromFFI(ptr, aEdges) {
+        const aDeref = /* TODO: gen_c_to_js_deref */null;
+        this.#a = aDeref(aEdges) // TODO: Slice c_to_js;
+        const bDeref = /* TODO: gen_c_to_js_deref */null;
+        this.#b = bDeref(aEdges) // TODO: Slice c_to_js;
+        const cDeref = /* TODO: gen_c_to_js_deref */null;
+        this.#c = cDeref(aEdges) // TODO: Slice c_to_js;
+
+        return this;
+    }
```
[the remark this answers is not in the document]
Comment (ambiguousname) [reply feature_tests/js2/api/BorrowedFields.mjs]: Also done!

File: tool/src/js2/formatter.rs (hunk 13 above)
Comment (ambiguousname) [reply tool/src/js2/formatter.rs:1]: Done!

File: tool/templates/js2/special_method.js.jinja
[the remark this answers is not in the document]
Comment (ambiguousname) [reply tool/templates/js2/special_method.js.jinja:1]: Done

**Comment by ambiguousname** [comment]
Since we're passing all checks, here's a list of everything that isn't done that was brought up during review.

Most of these I think can be iterated on in separate PRs:

- Is `BackendAttrSupport::memory_sharing` dead? Let's remove it if so.
- Lifetimes and edges
  - If we don't want to retain the `selfEdge` argument itself, we could `this.#selfEdge.push(...selfEdge)` (see `example/ICU4XDataProvider` for an example of this)
  - Are edges for structs, opaques, and slices implemented correctly? Do we clean all of these up correctly?
    - See `Foo.mjs`, `Bar.mjs`, and anything using `BorrowedFields` as an example.
  - Code passes tests, but garbage collection is not thoroughly tested.
- Helper functions in diplomatRuntime to minimize the amount of direct reference to `wasm.`
  - For instance, functions that return a DiplomatWrite buffer should do this.
- `==` vs. `===` for reading from buffers in JS. Is this desired behavior?
- Could we use `valueOf` and `[Symbol.toPrimitive]` for enumerators instead of `enum.value` or `enum.ffiValue`?
- Typescript tests for Javascript. Somewhat working, but require some fixes to the original JS backend. See #509 
- Separate .d.ts templates with shared structures like MethodInfo (to avoid generating the same information twice for each type)
- New typescript and javascript tests for additional functionality
  - Tests for borrowing types like `NestedBorrowedFields` and their lifetimes.
  - Tests for reading slices (a slice of strings especially) from function output
  - Tests for iterators and iterables (see next bullet point)
- Disabling all special method support statements in lib.rs, re-enable in separate PRs 
- Is there a way we can verify the argument positions of our functions?
  - I found that we can just read the LLVM IR output as a possible place to test outputs, but I think it'd be nice to find a place where we can explicitly see how a function in Rust corresponds to a function in WASM. If anyone wants to go digging with me into the Rust and LLVM compilers, let me know. 
- Are we supporting 'static  lifetimes for items like slices?

**Comment by ambiguousname** [comment]
If there's any changes we'd like to make to this PR let me know. I think I'd like to make a push to get this ready to merge into main and branching off these items into separate issues and PRs. But if there's anything we'd like to address in this PR, let me know!

File: feature_tests/js2/api/Foo.mjs [hunk 19]
```diff
@@ -8,12 +7,9 @@ import wasm from "./diplomat-wasm.mjs";
 import * as diplomatRuntime from "./diplomat-runtime.mjs";
 
 
-
-
 const Foo_box_destroy_registry = new FinalizationRegistry((ptr) => {
     wasm.Foo_destroy(ptr);
```
Comment (sffc) [inline feature_tests/js2/api/Foo.mjs:11]: Question: does this also need to destroy `aEdge` instances? I guess not; they are tracked on their own.

File: example/js2/api/ICU4XFixedDecimal.mjs [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    
+        try {
+    
+            if (!(result == 1)) {
+                 throw new diplomatRuntime.FFIError(null);}
+                return diplomatRuntime.readString8(wasm, wasm.diplomat_buffer_write_get_bytes(write), wasm.diplomat_buffer_write_len(write));
+        } finally {
+        
+            wasm.diplomat_buffer_write_destroy(write);
+
```
[the remark this answers is not in the document]
Comment (sffc) [reply example/js2/api/ICU4XFixedDecimal.mjs:68]: #522

File: example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+    // This struct contains borrowed fields, so this takes in a list of
+    // "edges" corresponding to where each lifetime's data may have been borrowed from
+    // and passes it down to individual fields containing the borrow.
+    // This method does not attempt to handle any dependencies between lifetimes, the caller
+    // should handle this when constructing edge arrays.
+    _fromFFI(ptr) {
+        const groupingStrategyDeref = diplomatRuntime.enumDiscriminant(wasm, ptr);
+        this.#groupingStrategy = ICU4XFixedDecimalGroupingStrategy[Array.from(ICU4XFixedDecimalGroupingStrategy.values.keys())[groupingStrategyDeref]];
+        const someOtherConfigDeref = (new Uint8Array(wasm.memory.buffer, ptr + 4, 1))[0] == 1;
```
[the remark this answers is not in the document]
Comment (sffc) [reply example/js2/api/ICU4XFixedDecimalFormatterOptions.mjs:40]: #523

File: example/js2/api/ICU4XFixedDecimalGroupingStrategy.mjs [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+        throw TypeError(value + " is not a ICU4XFixedDecimalGroupingStrategy and does not correspond to any of its enumerator values.");
+    }
+
+    get value() {
+        return this.#value;
+    }
+
+    get ffiValue() {
+        return ICU4XFixedDecimalGroupingStrategy.values.get(this.#value);
+    }
```
[the remark this answers is not in the document]
Comment (sffc) [reply example/js2/api/ICU4XFixedDecimalGroupingStrategy.mjs:35]: #524

File: feature_tests/js2/api/Bar.mjs [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+        
+        // This lifetime edge depends on lifetimes 'a
+        let aEdges = [this];
+        const result = wasm.Bar_foo(this.ffiValue);
+    
+        try {
+    
+            return new Foo(result, bEdges, aEdges);
+        } finally {
+
```
[the remark this answers is not in the document]
Comment (sffc) [reply feature_tests/js2/api/Bar.mjs:55]: #525

File: feature_tests/js2/api/Foo.mjs [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+        const result = wasm.Foo_as_returning(diplomat_receive_buffer, this.ffiValue);
+    
+        try {
+    
+            return new BorrowedFieldsReturning()._fromFFI(diplomat_receive_buffer, aEdges);
+        } finally {
+        
+            wasm.diplomat_free(diplomat_receive_buffer, 8, 4);
+        
+        }
```
[the remark this answers is not in the document]
Comment (sffc) [reply feature_tests/js2/api/Foo.mjs:108]: #525

File: feature_tests/js2/api/Float64Vec.mjs [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+    
+            return new Float64Vec(result, []);
+        } finally {
+        
+        }
+    }
+
+    get asBoxedSlice() {
+        
+        const diplomat_receive_buffer = wasm.diplomat_alloc(8, 4);
```
[the remark this answers is not in the document]
Comment (sffc) [reply feature_tests/js2/api/Float64Vec.mjs:136]: #522

File: feature_tests/js2/api/NestedBorrowedFields.mjs [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+
+
+export class NestedBorrowedFields {
+    #fields;
+    get fields()  {
+        return this.#fields;
+    }
+    set fields(value) {
+        this.#fields = value;
+    }
```
[the remark this answers is not in the document]
Comment (sffc) [reply feature_tests/js2/api/NestedBorrowedFields.mjs:16]: #526

File: feature_tests/js2/api/Foo.mjs (hunk 19 above)
Comment (sffc) [reply feature_tests/js2/api/Foo.mjs:11]: #525

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
Some follow-ups were filed; please open the remaining ones.

**Comment by ambiguousname** [comment]
Done!
